### PR TITLE
Introduce operations for syncing and rebasing

### PR DESCRIPTION
**Pull request by sstok**
|Q            |A   |
|---          |--- |
|Fixed Tickets| #410  |
|License      |MIT |

**This is a work in progress (extensive testing is required).**

Test:

Try all with need pull/push and diverged.

- [x] force
- [ ] smart with auto remote and branch
- [ ] smart with explicit remote and branch
- [ ] smart-merge
- [ ] smart with `--no-push`
- [ ] smart with auto remote and branch plus `--force-push`

The `branch:sync` command now supports three ways of synchronizing (both ways):

* FORCE: `git reset --hard remote/source-branch`
* SMART:  Determines what is best, rebase or push (default)
* SMART_MERGE_ONLY: Same as SMART but uses merge instead of rebase for pulling-in changes.

All except force ensure the local branch is updated (rebase or merge) with the remote branch and pushes when done (can be disabled).

Internally this uses the new `GitSyncOperation` class which can also be used for `pull-request:squash`, `pull-request:rebase` (coming soon) and `pull-request:switch-base`.

### CODE REVIEW

**Comment by sstok** [comment]
FYI the `WIP - BranchSyncCommand` commit needs to be cherry-picked in a separate PR.

File: src/Operation/GitSyncOperation.php [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+                break;
+        }
+    }
+
+    private function pushToRemote($force = false)
+    {
+        $target = trim($this->localBranch).':'.$this->destinationBranch;
+
+        // Safety guard to prevent deleting a remote base branch!!
+        if (':' === $target[0]) {
```
Comment (sstok) [inline src/Operation/GitSyncOperation.php]: I'm thinking about moving this safe-guard to pushToRemote() and then adding a new removeRemoteBranch() method for deleting remote branches :+1: Will do this in a separate PR with some other improvements.

File: src/Command/Branch/BranchSyncCommand.php [hunk 2]
```diff
@@ -28,18 +30,69 @@ protected function configure()
...
         $this
             ->setName('branch:sync')
-            ->setDescription('Syncs local branch with its upstream version')
-            ->addArgument('branch_name', InputArgument::OPTIONAL, 'Branch name to sync')
+            ->setDescription('Synchronises a local branch with a remote branch')
+            ->addArgument('source_branch', InputArgument::OPTIONAL, 'Remote branch-name to use as (new) base')
+            ->addArgument('source_remote', InputArgument::OPTIONAL, 'Remote to pull from (defaults to [your-username])')
             ->addArgument(
-                'remote',
+                'dest_remote',
```
Comment (pierredup) [inline src/Command/Branch/BranchSyncCommand.php]: This is a BC break. Can't you rather add both `remote` and `dest_remote` with a deprecation warning when using `remote`?

Comment (sstok) [inline src/Command/Branch/BranchSyncCommand.php]: The arguments are never executed with a name ;)
Options are, and left these unchanged (as much as possible).

So using `branch:sync sstok master` will still work :+1: 

One thing that is a BC break, is that instead of forcing the local branch to equal the remote
branch it uses rebase by default. But then again using rebase is saver then using `git reset` by default.

Comment (pierredup) [inline src/Command/Branch/BranchSyncCommand.php]: ah, I always get confused between options and arguments in the sf console :)

**Comment by sstok** [comment]
OK that went horrible :skull: 

```
CMD git rev-parse sstok/master
CMD git rebase --onto sstok/master 393467e8d85d6adf888fcdc3a2c46964720084cf new-test
CMD git push sstok --force new-test:master
```

**Comment by sstok** [comment]
I'm not completely sure on this solution anymore, instead I would favor on always using the local branch (so that it's up-to-date) BUT first check that the local branch is not outdated or has newer commits (which was the source of trouble in the past).

1. If there are newer commits, stop and require that the user either pushes it's local branch or does a hard reset (don't this automatically! some people forget to create a new branch for working)
2. If the local branch is outdated, pull-in the changes.
3. When a diverged state is encountered stop (see nr. 1)

This will make the logic easier to understand, prevents having temp branches (that were not removed) and removes the need for some hacks in https://github.com/gushphp/gush/blob/master/src/Helper/GitHelper.php#L437-L442

**Comment by sstok** [comment]
Closing this one in favor a smaller pull-request (with the whole sync operation stuff).
